### PR TITLE
Support Ruby 2.4

### LINKED ISSUE #820: Fails to build under Ruby 2.4.0.
`backup` has a firm requirement on `json` version `1.8.2`, which fails to build with Ruby 2.4.0 due to `Fixnum` and `Bignum` being merged into `Integer`.

```shell
current directory: /usr/local/lib/ruby/gems/2.4.0/gems/json-1.8.2/ext/json/ext/generator
make "DESTDIR="
compiling generator.c
generator.c: In function ‘generate_json’:
generator.c:867:25: error: ‘rb_cFixnum’ undeclared (first use in this function)
     } else if (klass == rb_cFixnum) {
                         ^
generator.c:867:25: note: each undeclared identifier is reported only once for each function it appears in
generator.c:869:25: error: ‘rb_cBignum’ undeclared (first use in this function)
     } else if (klass == rb_cBignum) {
```

Updating to `json` version `2.0.2` corrects this problem. Please update the `gemspec` when you get a chance, thanks!

### PR DESCRIPTION
**Pull request by tombruijn**
Closes #820 

Currently doesn't bundle because of qiniu gem relies on json ~> 1.8

### CODE REVIEW

**Comment by tombruijn** [comment]
PR created: https://github.com/qiniu/ruby-sdk/pull/183

File: Gemfile [hunk 1]
```diff
@@ -34,7 +34,7 @@ group :production do
   gem 'twitter'
   gem 'hipchat'
   gem 'flowdock'
-  gem 'json'
+  gem 'json', '2.0.2'
```
Comment (knu) [inline Gemfile]: You shouldn't need to hardcode a specific version of json any more.  The json gem 1.8.6 and 2.0.3 both support all versions of ruby.  2.0.3 now only requires ruby >=1.9, and 1.8.6 supports ruby >=2.0.

**Comment by tombruijn** [comment]
Updated PR to be aimed against the new `master` branch. The dependency solution in the v4 branch makes it difficult to support old versions and Ruby 2.4 at the same time.
Unless someone wants to pick up that problem in the v4 branch, the new major v5 version will only support Ruby 2.4

**Comment by nynhex** [comment]
@tombruijn Seems like the v4 branch backporting of 2.4 may be a bit of a chore.  Personally I don't see an issue with 2.4.x users being encouraged to use the new v5 branch.  Also when v5 is cut let me know as I'm going to port Dropbox V2 API into it and will make a pass at v4 stable to ensure the Dropbox API V1 deprecation doesn't affect users.

**Comment by tombruijn** [comment]
I don't think we'll do a lot of changes in v5 anymore. I'd rather release something soon than wait spend forever on a new version. It's not a major change in terms of logic that's has changed, but mainly a lot of code polish, which is still on going. I don't think we'll get the entire milestone done: https://github.com/backup/backup/milestone/1
but maybe _some_ integration testing merged before v5 is released?

**Comment by nynhex** [comment]
I think that's reasonable.  I do fancy the goal on the milestone of switching to Docker.  I can help on that if you need it.

**Comment by tombruijn** [comment]
@nynhex Great to see you're willing to help on that :) I've left some thoughts here: https://github.com/backup/backup/pull/832#issuecomment-294368266

**Pull request merged**